File: types/nmea-baud-rate.go
package types

// NMEABaudRate represents the expected baud rate the sensor uses to decode
// NMEA UART input $GPRMC messages.
// For additional information, refer to [nmea_baud_rate].
//
// [nmea_baud_rate]: https://static.ouster.dev/sensor-docs/image_route1/image_route2/common_sections/API/sensor_configuration_description.html#nmea-baud-rate
type NMEABaudRate int

const (
	NMEABaudRateUnknown NMEABaudRate = iota // An unknown or undefined baud rate.
	NMEABaudRate9600                        // 9600 baud for NMEA UART.
	NMEABaudRate115200                      // 115200 baud for NMEA UART.
)

// String returns the string representation of a NMEABaudRate value.
func (r NMEABaudRate) String() string {
	switch r {
	default:
		fallthrough
	case NMEABaudRateUnknown:
		return "unknown"
	case NMEABaudRate9600:
		return "9600 baud"
	case NMEABaudRate115200:
		return "115200 baud"
	}
}

// GoString returns the Go syntax representation of a NMEABaudRate value.
func (r NMEABaudRate) GoString() string {
	switch r {
	default:
		fallthrough
	case NMEABaudRateUnknown:
		return "NMEABaudRateUnknown"
	case NMEABaudRate9600:
		return "NMEABaudRate9600"
	case NMEABaudRate115200:
		return "NMEABaudRate115200"
	}
}

// MarshalText returns the text representation of a NMEABaudRate value.
func (r NMEABaudRate) MarshalText() ([]byte, error) {
	switch r {
	default:
		fallthrough
	case NMEABaudRateUnknown:
		return []byte("UNKNOWN"), nil
	case NMEABaudRate9600:
		return []byte("BAUD_9600"), nil
	case NMEABaudRate115200:
		return []byte("BAUD_115200"), nil
	}
}

// UnmarshalText updates the value of NMEABaudRate based on the input text.
// It maps the input text to the corresponding NMEABaudRate value.
// If the input text is not recognized, NMEABaudRateUnknown is set.
func (r *NMEABaudRate) UnmarshalText(text []byte) error {
	switch string(text) {
	default:
		*r = NMEABaudRateUnknown
	case "BAUD_9600":
		*r = NMEABaudRate9600
	case "BAUD_115200":
		*r = NMEABaudRate115200
	}

	return nil
}

File: src/types/nmea_baud_rate.rs
use num_derive::{FromPrimitive, ToPrimitive};
use std::fmt::{Display, Formatter};

/// Represents the expected baud rate the sensor uses to decode NMEA UART input `$GPRMC` messages.
///
/// For additional information, refer to [`nmea_baud_rate`].
///
/// [`nmea_baud_rate`]: https://static.ouster.dev/sensor-docs/image_route1/image_route2/common_sections/API/sensor_configuration_description.html#nmea-baud-rate
#[derive(
    Debug,
    serde::Serialize,
    serde::Deserialize,
    Eq,
    PartialEq,
    Copy,
    Clone,
    FromPrimitive,
    ToPrimitive,
)]
#[repr(u32)]
pub enum NmeaBaudRate {
    /// 9600 baud.
    #[serde(rename = "BAUD_9600")]
    Baud9600 = 9600,

    /// 115200 baud.
    #[serde(rename = "BAUD_115200")]
    Baud115200 = 115200,
}

impl Display for NmeaBaudRate {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{} baud",
            match self {
                NmeaBaudRate::Baud9600 => "9600 baud",
                NmeaBaudRate::Baud115200 => "115200 baud",
            }
        )
    }
}

impl Default for NmeaBaudRate {
    fn default() -> Self {
        Self::Baud9600
    }
}
